README.md
# wefarm-task

## Lachlan Hespe

A Clojure library to turn an integer input into a natural language string eg. 99 -> 'ninety nine'

## Technical Decisions

- Keep the code simple and readable.
- Take an iterative approach to solving the problem. Tests preceeded most of the code  to get something working. Some of the tests are now redundant, but show the progression through the code. (also see the commit history for more detail)
- Follow consistent clojure style as much as possible, used [https://guide.clojure.style].

## Notes

- Maybe some more examples of natural language numbers. I used 3007002 as a test and had to think a bit whether "three million _and_ seven thousand and two" may be correct, decided it wasn't :).
- I feel that the my implementation could be further collapsed using more functional language features, but had to decide when to stop, I look forward to getting feedback on how it could be improved.

## Build

Commands listed here require having lein/clojure/java installed and accessible on the path [*](#Docker-alternative).

```bash
lein jar
```

## Test

```bash
lein test
```

### Coverage

```bash
lein cloverage
```

## Usage

```clojure
(num->word 99) ;=> returns ninety nine
```

#### Docker alternative

If you don't have lein installed it is possible to run the lein commands using the standard clojure docker image. Use `docker run -it --rm -v "$PWD":/usr/src/app -w /usr/src/app clojure lein cloverage` any artifacts of the lein command will probably have permissions matching the docker daemon (everything was created as root when tested locally), it should be possible to overcome this with the `-u` flag in docker but that is tangential to the task at hand.


## License

Copyright © 2021 Lachlan Hespe

Distributed under the Eclipse Public License either version 1.0 or (at your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2021-02-26
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2021-02-26
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/wefarm-task/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/wefarm-task/compare/0.1.0...0.1.1


test/wefarm_task/core_test.clj
(ns wefarm-task.core-test
  (:require [clojure.test :refer :all]
            [wefarm-task.core :refer :all]))

(deftest bootstrap-test
  (testing "bootstrap test, do something"
    (is (= (num->word 42) "forty two"))))

(deftest input-bounds-test
  (testing "check that out of bounds input fails"
    (is (thrown? java.lang.AssertionError (num->word -500)))
    (is (thrown? java.lang.AssertionError (num->word "99")))
    (is (thrown? java.lang.AssertionError (num->word 1000000000)))))

(deftest ones-test
  (testing "values 0 through 9"
    (is (= (num->word 0) "zero"))
    (is (= (num->word 1) "one"))
    (is (= (num->word 9) "nine"))))

(deftest teens-test
  (testing "values 10 through 19"
    (is (= (num->word 10) "ten"))
    (is (= (num->word 19) "nineteen"))))

(deftest tens-test
  (testing "values 20 through 99"
    (is (= (num->word 20) "twenty"))
    (is (= (num->word 26) "twenty six"))
    (is (= (num->word 99) "ninety nine"))
    (is (= (num->word 42) "forty two"))))

(deftest hundreds-test
  (testing "values 100 through 999"
    (is (= (num->word 100) "one hundred"))
    (is (= (num->word 102) "one hundred and two"))
    (is (= (num->word 116) "one hundred and sixteen"))))

(deftest thousands-test
  (testing "values 1000 through 9999"
    (is (= (num->word 1000) "one thousand"))
    (is (= (num->word 1002) "one thousand and two"))
    (is (= (num->word 1016) "one thousand and sixteen"))
    (is (= (num->word 1020) "one thousand and twenty"))
    (is (= (num->word 1120) "one thousand one hundred and twenty"))))

(deftest ten-thousands-test
  (testing "values 10,000 - 99,000"
    (is (= (num->word 10000) "ten thousand"))
    (is (= (num->word 10005) "ten thousand and five"))
    (is (= (num->word 10105) "ten thousand one hundred and five"))))

(deftest hundred-thousands-test
  (testing "values 100,000 - 999,999"
    (is (= (num->word 100000) "one hundred thousand"))
    (is (= (num->word 110005) "one hundred and ten thousand and five"))
    (is (= (num->word 917055) "nine hundred and seventeen thousand and fifty five"))
    (is (= (num->word 207255) "two hundred and seven thousand two hundred and fifty five"))))

(deftest millions-test
  (testing "values 1,000,000 - 9,999,999"
    (is (= (num->word 1000000) "one million"))
    (is (= (num->word 9517000) "nine million five hundred and seventeen thousand"))
    (is (= (num->word 3007002) "three million seven thousand and two"))))

(deftest ten-millions-test
  (testing "values 10,000,000 - 99,999,999"
    (is (= (num->word 10000000) "ten million"))
    (is (= (num->word 17517000) "seventeen million five hundred and seventeen thousand"))
    (is (= (num->word 99007002) "ninety nine million seven thousand and two"))))

(deftest hundred-millions-test
  (testing "values 100,000,000 - 999,999,999"
    (is (= (num->word 100000000) "one hundred million"))
    (is (= (num->word 311517000) "three hundred and eleven million five hundred and seventeen thousand"))
    (is (= (num->word 300017000) "three hundred million seventeen thousand"))
    (is (= (num->word 999999999) "nine hundred and ninety nine million nine hundred and ninety nine thousand nine hundred and ninety nine"))))

(deftest task-tests
  (testing "the values specified in the task"
    (is (= (num->word 0) "zero"))
    (is (= (num->word 1) "one"))
    (is (= (num->word 21) "twenty one"))
    (is (= (num->word 105) "one hundred and five"))
    (is (= (num->word 123) "one hundred and twenty three"))
    (is (= (num->word 1005) "one thousand and five"))
    (is (= (num->word 1042) "one thousand and forty two"))
    (is (= (num->word 1105) "one thousand one hundred and five"))
    (is (= (num->word 56945781) "fifty six million nine hundred and forty five thousand seven hundred and eighty one"))
    (is (= (num->word 999999999) "nine hundred and ninety nine million nine hundred and ninety nine thousand nine hundred and ninety nine"))))

(deftest big-number
  (testing "a number bigger than a billion"
    (is (= (num->word 1000000001) "one billion and one"))))


src/wefarm_task/core.clj
(ns wefarm-task.core
  (:require [clojure.core :refer :all]
            [clojure.string :as str]))

(def num-names {:ones ["zero" "one" "two" "three" "four"
                       "five" "six" "seven" "eight" "nine"]
                :teens ["ten" "eleven" "twelve" "thirteen" "fourteen"
                        "fifteen" "sixteen" "seventeen" "eighteen" "nineteen"]
                :tens [nil nil "twenty" "thirty" "forty"
                       "fifty" "sixty" "seventy" "eighty" "ninety"]
                })

(def period-names [nil "thousand" "million" "billion" "trillion"])

(def min-int 0)
;; easier notation for expressing max billion 1e9, trillion 1e12
(def max-int (dec (bigint 1e12)))

(defn- valid-number?
  [x]
  (and (integer? x) (<= min-int x max-int)))

(defn- digits
  "Transforms an integer to a reversed vector of digits.
  eg. 123 -> [3 2 1]"
  [x]
  (->> x
       (iterate #(quot % 10))
       (take-while pos?)
       (mapv #(mod % 10))))

(defn- ones
  [d]
  ((:ones num-names) d))

(defn- tens
  [ds]
  (str
   ((:tens num-names) (last ds))
   (if (pos? (first ds))
     (str " " (ones (first ds))))))

(defn- tens-ones
  "Checks if digit sequence passed in is a ones 0-9,
  teens 10-19 or tens 19-99 and maps to words."
  [ds]
  (let [has-tens (and (= (count ds) 2) (pos? (last ds)))
        is-teen (and has-tens (= (last ds) 1))]
    (if has-tens
      (if is-teen
        ((:teens num-names) (first ds))
        (tens ds))
      (ones (first ds)))))

(defn- tens-ones-prefix
  "Adds 'and' to tens and ones that need it to satisfy
  natural language rules 1021 - one thousand and twenty one"
  [ds prefix]
  (let [tens (tens-ones ds)
        has-tens (pos? (reduce + ds))]
    (if has-tens
      (if prefix
        (str "and " tens)
        tens))))

(defn- hundreds
  "Transforms the digit sequence to hundreds, tens and ones. If
  hundreds are present overrides prefix to true otherwise passes
  it along (this controls inclusion of 'and')."
  [ds prefix]
  (let [has-hundreds (and (= (count ds) 3) (pos? (last ds)))
        hundreds (str (ones (last ds)) " " "hundred")
        tens-pos (if (= (count ds) 3) (pop ds) ds)]
    (if has-hundreds
      (str/join " " (keep identity [hundreds (tens-ones-prefix tens-pos true)]))
      (tens-ones-prefix tens-pos prefix))))

(defn- period-name
  [i]
  (if-let [p (period-names i)]
    (str " " p)))

(defn- period
  "Takes a period (number part grouped in 3's) the index of the part and
  the length of the overall number. Returns the natural language version
  of the period with the period name added (thousand, million etc)"
  [ds i len]
  (if-let [period (hundreds ds (and (= i 0) (> len 3)))]
    (str period (period-name i))))

(defn num->word
  "Takes a integer between 0 and 999999999 and returns the natural
  language version eg. 123 -> 'one hundred and twenty three'."
  [x]
  {:pre [(valid-number? x)]}
  (let [ds (digits x)
        len (count ds)
        periods (map vec (partition 3 3 [] ds))]
    (if (= x 0)
      (ones x); special case for zero
      (->> periods
           (map-indexed #(period %2 %1 len))
           (keep identity) ; remove nils
           reverse
           (str/join " ")))))
